apps/backend/src/action/logic/check-tool-has-broken.ts

```diff
@@ -1,27 +1,133 @@
+import { Eq as stringEq } from "fp-ts/string";
+import * as E from "fp-ts/Either";
 import { BrokenToolCommand } from "~/action/command";
-import { ResultAsync, errAsync } from "neverthrow";
+import {
+  BrokenToolHasBeenPlacedEvent,
+  BrokenToolHasBeenRemovedEvent,
+  Event,
+  isBrokenToolHasBeenPlacedEvent,
+  isBrokenToolHasBeenRemovedEvent,
+} from "~/action/event";
+import type { EventSource } from "~/models/event";
+import { PlayerToolState, Tool, ToolState } from "~/models/tool";
+import { always, assoc, error } from "~/utils";
+
+export const InitialPlayerToolState: PlayerToolState = {
+  [Tool.Cart]: ToolState.NotBroken,
+  [Tool.Lamp]: ToolState.NotBroken,
+  [Tool.Pickaxe]: ToolState.NotBroken,
+};
+
+const EventSourceReadError = error("EventSourceReadError");
+const ToolHasBeenBrokenError = error("ToolHasBeenBrokenError");
+const ToolHasNotBeenBrokenError = error("ToolHasNotBeenBrokenError");
+
+type EventSourceReadError = ReturnType<typeof EventSourceReadError>;
+type ToolHasBeenBrokenError = ReturnType<typeof ToolHasBeenBrokenError>;
+type ToolHasNotBeenBrokenError = ReturnType<typeof ToolHasNotBeenBrokenError>;
+
+export type CheckToolHasBrokenError =
+  | EventSourceReadError
+  | ToolHasBeenBrokenError
+  | ToolHasNotBeenBrokenError;
+
+type PlayerToolStateOrError = E.Either<
+  CheckToolHasBrokenError,
+  PlayerToolState
+>;
 
 export interface CheckToolHasBroken {
   (repository: EventSource<Event>, command: BrokenToolCommand): ResultAsync<
     boolean,
-    Error
+    CheckToolHasBrokenError
   >;
 }
 
+const readAllEventsFromEventSource = (source: EventSource<Event>) =>
+  ResultAsync.fromPromise(
+    source.read(),
+    always(EventSourceReadError("failed to read events from source"))
+  );
+
+const filterAllEventsByPlayerId =
+  (command: BrokenToolCommand) => (event: Event) =>
+    match(event.data)
+      .with({ playerId: command.data.playerId }, always(true))
+      .otherwise(always(false));
+
+const validatePlayerToolCanBeBroken =
+  (stateOrError: PlayerToolStateOrError) =>
+  ({ data: { playerId, tool } }: BrokenToolHasBeenPlacedEvent) =>
+    pipe(
+      stateOrError,
+      E.chain((state) =>
+        state[tool] === ToolState.NotBroken
+          ? E.right(assoc(tool, ToolState.Broken, state))
+          : E.left(
+              ToolHasBeenBrokenError(
+                `can not broke player ${playerId} tool ${tool}`
+              )
+            )
+      )
+    );
+
+const validatePlayerToolCanBeFixed =
+  (stateOrError: PlayerToolStateOrError) =>
+  ({ data: { playerId, tool } }: BrokenToolHasBeenRemovedEvent) =>
+    pipe(
+      stateOrError,
+      E.chain((state) =>
+        state[tool] === ToolState.Broken
+          ? E.right(assoc(tool, ToolState.NotBroken, state))
+          : E.left(
+              ToolHasNotBeenBrokenError(
+                `can not fix player ${playerId} tool ${tool}`
+              )
+            )
+      )
+    );
+
+const updatePlayerToolState = (
```

Comment:
I prefer using `array` in this case, 

- when we match `isBrokenToolHasBeenPlacedEvent`, 
   we can append the tool to the array. 
- when we match `isBrokenToolHasBeenRemovedEvent`, 
   we filter out the tool in the array.
   
 Trying to avoid mutating will let us match easier to tracing bugs.

Reply:
That's a great approach